File: modules/video-frame-decoder/ios/VideoFrameDecoderModule.swift
import ExpoModulesCore
import ExpoUI

public class VideoFrameDecoderModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VideoFrameDecoder")

    Events("onChange")

    Constant("PI") {
      Double.pi
    }

    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    View(VideoFrameDecoderView.self) {
      Events("onTap")
    }

    Class(VideoFrameDecoderModuleSharedObject.self) {
      Constructor { () -> VideoFrameDecoderModuleSharedObject in
        return VideoFrameDecoderModuleSharedObject()
      }

      Property("count") { (ref: VideoFrameDecoderModuleSharedObject) -> Int in
        return ref.count
      }
      .set { (ref: VideoFrameDecoderModuleSharedObject, count: Int) in
        ref.count = count
      }
    }

    ExpoUIView(VideoFrameDecoderSwiftUIView.self)

    OnCreate {
      ViewModifierRegistry.register("videoFrameDecoderSwiftUIModifier") { params, appContext, _ in
        return try VideoFrameDecoderSwiftUIModifier(from: params, appContext: appContext)
      }
    }

    OnDestroy {
      ViewModifierRegistry.unregister("videoFrameDecoderSwiftUIModifier")
    }
  }
}

File: modules/video-frame-decoder/android/src/main/java/expo/modules/videoframedecoder/VideoFrameDecoderModule.kt
package expo.modules.videoframedecoder

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.ui.ExpoUIView
import expo.modules.kotlin.records.recordFromMap
import expo.modules.ui.ModifierRegistry

class VideoFrameDecoderModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("VideoFrameDecoder")

    Events("onChange")

    Constant("PI") {
      Math.PI
    }

    Function("hello") {
      "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { value: String ->
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    View(VideoFrameDecoderView::class) {
      // Defines an event that the view can send to JavaScript.
      Events("onTap")
    }

    Class(VideoFrameDecoderModuleSharedObject::class) {
      Constructor {
        val instance = VideoFrameDecoderModuleSharedObject(appContext)
        return@Constructor instance
      }

      Property("count")
        .get { ref: VideoFrameDecoderModuleSharedObject ->
          ref.count
        }
        .set { ref: VideoFrameDecoderModuleSharedObject, count: Int ->
          ref.count = count
        }
    }

    ExpoUIView<VideoFrameDecoderComposeViewProps>("VideoFrameDecoderComposeView") {
      Content { props ->
        VideoFrameDecoderComposeViewContent(props)
      }
    }

    OnCreate {
      ModifierRegistry.register("videoFrameDecoderComposeModifier") { params, _, _, _ ->
        recordFromMap<VideoFrameDecoderComposeModifierParams>(params).toModifier()
      }
    }
  }
}
